libs/vkd3d/device.c

```diff
@@ -2234,17 +2293,41 @@ static void d3d12_device_destroy(struct d3d12_device *device)
 static ULONG STDMETHODCALLTYPE d3d12_device_Release(d3d12_device_iface *iface)
 {
     struct d3d12_device *device = impl_from_ID3D12Device(iface);
-    ULONG refcount = InterlockedDecrement(&device->refcount);
+    ULONG cur_refcount, cas_refcount;
+    bool is_locked = false;
 
-    TRACE("%p decreasing refcount to %u.\n", device, refcount);
+    cur_refcount = 0;
+    cas_refcount = vkd3d_atomic_uint32_load_explicit(&device->refcount, vkd3d_memory_order_relaxed);
 
-    if (!refcount)
+    /* In order to prevent another thread from resurrecting a destroyed device,
+     * we need to lock the container mutex before decrementing the ref count to
+     * zero, so we'll have to use a CAS loop rather than basic atomic decrement */
+    while (cas_refcount != cur_refcount)
     {
+        cur_refcount = cas_refcount;
+
+        if (cur_refcount == 1 && !is_locked)
+        {
+            pthread_mutex_lock(&d3d12_device_map_mutex);
+            is_locked = true;
+        }
+
+        cas_refcount = vkd3d_atomic_uint32_compare_exchange((uint32_t*)&device->refcount, cur_refcount,
+                cur_refcount - 1, vkd3d_memory_order_acquire_release, vkd3d_memory_order_relaxed);
```

Comment:
`vkd3d_memory_order_acquire_release` → `vkd3d_memory_order_acq_rel`, otherwise the build fails with

```
../vkd3d-proton/libs/vkd3d/device.c: In function ‘d3d12_device_Release’:
../vkd3d-proton/libs/vkd3d/device.c:2316:35: error: ‘vkd3d_memory_order_acquire_release’ undeclared (first use in this function); did you mean ‘vkd3d_memory_order_acq_rel’?
 2316 |                 cur_refcount - 1, vkd3d_memory_order_acquire_release, vkd3d_memory_order_relaxed);
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                   vkd3d_memory_order_acq_rel
```

Reply:
^

Reply:
fixed.